Purchases/SwiftSources/LocalReceiptParsing/Builders/ASN1ContainerBuilder.swift

```diff
@@ -0,0 +1,96 @@
+        var internalContainers: [ASN1Container] = []
+        if encodingType == .constructed {
+            internalContainers = try buildInternalContainers(payload: internalPayload)
+        }
+        return ASN1Container(containerClass: containerClass,
+                             containerIdentifier: containerIdentifier,
+                             encodingType: encodingType,
+                             length: length,
+                             internalPayload: internalPayload,
+                             internalContainers: internalContainers)
+    }
+}
+
+private extension ASN1ContainerBuilder {
+
+    func buildInternalContainers(payload: ArraySlice<UInt8>) throws -> [ASN1Container] {
+        var internalContainers = [ASN1Container]()
+        var currentPayload = payload
+        while (currentPayload.count > 0) {
+            let internalContainer = try build(fromPayload: currentPayload)
+            internalContainers.append(internalContainer)
+            currentPayload = currentPayload.dropFirst(internalContainer.totalBytesUsed)
+        }
+        return internalContainers
+    }
+
+    func extractClass(byte: UInt8) throws -> ASN1Class {
+        let firstTwoBits = byte.valueInRange(from: 0, to: 1)
+        guard let asn1Class = ASN1Class(rawValue: firstTwoBits) else {
+            throw ReceiptReadingError.asn1ParsingError(description: "couldn't determine asn1 class")
+        }
+        return asn1Class
+    }
+
+    func extractEncodingType(byte: UInt8) throws -> ASN1EncodingType {
+        let thirdBit = byte.bitAtIndex(2)
+        guard let encodingType = ASN1EncodingType(rawValue: thirdBit) else {
+            throw ReceiptReadingError.asn1ParsingError(description: "couldn't determine encoding type")
+        }
+        return encodingType
+    }
+
+    func extractIdentifier(byte: UInt8) throws -> ASN1Identifier {
+        let lastFiveBits = byte.valueInRange(from: 3, to: 7)
+        guard let asn1Identifier = ASN1Identifier(rawValue: lastFiveBits) else {
+            throw ReceiptReadingError.asn1ParsingError(description: "couldn't determine identifier")
+        }
+        return asn1Identifier
+    }
+
+    func extractLength(data: ArraySlice<UInt8>) throws -> ASN1Length {
+        guard let firstByte = data.first else {
+            throw ReceiptReadingError.asn1ParsingError(description: "length needs to be at least one byte")
+        }
+
+        let lengthBit = firstByte.bitAtIndex(0)
+        let isShortLength = lengthBit == 0
+
+        let firstByteValue = Int(firstByte.valueInRange(from: 1, to: 7))
+
+        if isShortLength {
+            return ASN1Length(value: firstByteValue, bytesUsedForLength: 1)
+        } else {
+            let totalLengthOctets = firstByteValue
+            let lengthBytes = data.dropFirst().prefix(totalLengthOctets)
+            let lengthValue = lengthBytes.toInt()
+            return ASN1Length(value: lengthValue, bytesUsedForLength: totalLengthOctets + 1)
```

Comment:
Why are you summing 1 to the `totalLengthOctets`? Isn't `totalLengthOctets` the total already?

Reply:
well, `totalLenghtOctets` is the number of octets (I'll rename to bytes for consistency by the way) that the length header indicates you need to read to find the length. 
i.e.: if a length is 3 bytes long, you get 4 bytes: 
1st a header with the first bit set to 1 (indicating long length) + the value "3"
then 3 bytes with the actual value

Reply:
I did a small refactor here, I hope it clears things up a bit